### PR TITLE
DDF-4334 Add Support for Federated Sources in Table Export

### PR DESCRIPTION
**Pull request by willwill96**
#### What does this PR do?
1. Adds ability for Cql Requests to query against multiple sources
2. Enables Table Export to Query against federated sources

#### Who is reviewing it? 
@rymach 
@mojogitoverhere 
@GabrielFabian 
@gjvera 

#### Select relevant component teams: 
@codice/ui 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@djblue

#### How should this be tested?
1. Set up both local catalog and at least one federated source (each with multiple resources)
2. Run searches against multiple combinations of sources
	- Ensure that results are the ones expected
	- Export results using table export and ensure that all metacards in the table are exported
3. Spot test search functionality for regressions
#### What are the relevant tickets?
[DDF-4334](https://codice.atlassian.net/browse/DDF-4334)
#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 1]
```diff
@@ -26,9 +26,7 @@
 import ddf.catalog.operation.QueryRequest;
 import ddf.catalog.operation.impl.QueryImpl;
 import ddf.catalog.operation.impl.QueryRequestImpl;
-import java.util.Collections;
-import java.util.List;
-import java.util.Locale;
+import java.util.*;
```
Comment (willwill96) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: Remove * import

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js [hunk 2]
```diff
@@ -192,7 +192,7 @@ Query.Model = Backbone.AssociatedModel.extend({
     var sources = data.src
     var initialStatus = sources.map(function(src) {
       return {
-        id: src,
+        id: "[" + [src].toString().replace(",", ", ") + "]",
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js]: Remove replace as it doesn't do anything here

File: ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js [hunk 3]
```diff
@@ -248,7 +248,7 @@ Query.Model = Backbone.AssociatedModel.extend({
     }
     var query = this
     this.currentSearches = sources.map(function(src) {
-      data.src = src
+      data.srcs = [src]
```
Comment (willwill96) [inline ui/packages/catalog-ui-search/src/main/webapp/js/model/Query.js]: Check for other places that may need to be updated to send an array of strings instead of a single string

**Comment by willwill96** [comment]
Still need to add more backend tests to test the new functionality

**Comment by willwill96** [comment]
@andrewkfiedler In that case, would you say that src or srcs should be defaulted to when making the query (in the edge case that the query contains both)

File: distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java [hunk 4]
```diff
@@ -1465,13 +1465,13 @@ public void testMetacardCache() throws Exception {
...
-        "{\"src\":\""
+        "{\"srcs\": [\""
             + OPENSEARCH_SOURCE_ID
-            + "\",\"start\":1,\"count\":250,\"cql\":\"anyText ILIKE '*'\",\"sort\":\"modified:desc\"}";
+            + "\"],\"start\":1,\"count\":250,\"cql\":\"anyText ILIKE '*'\",\"sort\":\"modified:desc\"}";
 
     expect("Waiting for metacard cache to clear")
         .checkEvery(1, SECONDS)
-        .within(20, SECONDS)
+        .within(65, SECONDS)
```
Comment (willwill96) [inline distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java]: Change this back

**Comment by gjvera** [comment]
Looks like you need a rebase or something? You have a commit that's from another PR. Is that intentional?

**Comment by willwill96** [comment]
There was a bug with the jacoco plugin. I needed to pull in the changes from [this PR](https://github.com/codice/ddf/pull/4040) to figure out what they needed to be changed to. Once it's merged, I'll rebase.

File: catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/cql/CqlRequestTest.java [hunk 5]
```diff
@@ -152,6 +154,28 @@ public void testBlankLocalSource() {
...
   }
 
+  @Test
+  public void testMultipleSources() {
+    String[] sources = {"source1", "source2"};
+    cqlRequest.setSrcs(sources);
+    cqlRequest.createQueryRequest("source1", filterBuilder);
+    assertThat(Arrays.asList(cqlRequest.getSrcs()), contains("source1", "source2"));
+  }
+
```
Comment (cantstoptheunk) [inline catalog/ui/catalog-ui-search/src/test/java/org/codice/ddf/catalog/ui/query/cql/CqlRequestTest.java]: Might be worth considering adding a test case where "source1" is all-caps like `SOURCE1`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 6]
```diff
@@ -159,15 +172,20 @@ public QueryRequest createQueryRequest(String localSource, FilterBuilder filterB
...
-    String source = parseSrc(localSource);
-
     QueryRequest queryRequest;
-    if (CACHE_SOURCE.equals(source)) {
-      queryRequest = new QueryRequestImpl(query, true);
-      queryRequest.getProperties().put("mode", CACHE_SOURCE);
+    if (srcs != null && srcs.length != 0) {
+      parseSrcs(localSource);
+      queryRequest = new QueryRequestImpl(query, Arrays.asList(srcs));
+      queryRequest.getProperties().put(MODE, "update");
```
Comment (hayleynorton) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ✏️ why not make `"update"` a constant too?

**Comment by willwill96** [comment]
build now

**Comment by willwill96** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 7]
```diff
@@ -159,15 +174,20 @@ public QueryRequest createQueryRequest(String localSource, FilterBuilder filterB
...
 
-    String source = parseSrc(localSource);
-
     QueryRequest queryRequest;
-    if (CACHE_SOURCE.equals(source)) {
-      queryRequest = new QueryRequestImpl(query, true);
-      queryRequest.getProperties().put("mode", CACHE_SOURCE);
+    if (srcs != null && srcs.length != 0) {
+      parseSrcs(localSource);
+      queryRequest = new QueryRequestImpl(query, Arrays.asList(srcs));
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ✏️ `Collections.singletonList`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 8]
```diff
@@ -203,6 +223,22 @@ private String parseSrc(String localSource) {
     return src;
   }
 
+  private void parseSrcs(String localSource) {
+    for (int i = 0; i < srcs.length; i++) {
+      if (StringUtils.equalsIgnoreCase(srcs[i], LOCAL_SOURCE) || StringUtils.isBlank(srcs[i])) {
```
Comment (bdeining) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: in what situations will `srcs` have blank strings?

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java (hunk 7 above)
Comment (willwill96) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: In this `if` block `srcs` will likely have more than one string contained in it.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java (hunk 8 above)
Comment (willwill96) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: I extrapolated this function from the `parseSrc` function that was already present for a single source and this is the check that was used there. I suppose there wouldn't in practice be a case where `srcs`has a blank string - I will remove the `isBlank` portion

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 9]
```diff
@@ -71,6 +78,10 @@
 
   private boolean excludeUnnecessaryAttributes = true;
 
+  public String[] getSrcs() {
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ✏️ Prefer Collections over arrays.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 10]
```diff
@@ -203,6 +223,22 @@ private String parseSrc(String localSource) {
...
+    for (int i = 0; i < srcs.length; i++) {
+      if (StringUtils.equalsIgnoreCase(srcs[i], LOCAL_SOURCE) || StringUtils.isBlank(srcs[i])) {
+        srcs[i] = localSource;
+      }
+    }
+  }
+
+  public String getSourceResponseString() {
+    if (srcs != null && srcs.length != 0) {
+      return Arrays.toString(srcs);
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ❗️ `Arrays.toString` string representation is not defined.  

https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#toString-java.lang.Object:A-

Use a collection and `String.join` to control the format.

https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#join-java.lang.CharSequence-java.lang.Iterable-

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 11]
```diff
@@ -95,6 +106,10 @@ public void setSrc(String src) {
     this.src = src;
   }
 
+  public void setSrcs(List<String> srcs) {
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ✏️ It would be less confusing now if we merged `src` and `srcs`.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 12]
```diff
@@ -203,6 +223,22 @@ private String parseSrc(String localSource) {
     return src;
   }
 
+  private void parseSrcs(String localSource) {
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ✏️ Since `parseSrc` and `parseSrcs` both have side effects on the fields, it would be clearer to move this logic into their setters.

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java [hunk 13]
```diff
@@ -203,6 +223,22 @@ private String parseSrc(String localSource) {
...
+    for (int i = 0; i < srcs.size(); i++) {
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: ✏️ This method could be replaced with `Collections.replaceAll`.

https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#replaceAll-java.util.List-T-T-

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java (hunk 11 above)
Comment (willwill96) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: This was what I originally did. @andrewkfiedler suggested supporting both so we don't have unwanted side effects related to the existing front end code that uses `src`

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java (hunk 13 above)
Comment (willwill96) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: Since we are doing case insensitive comparison, it doesn't look like this will work

**Comment by willwill96** [comment]
build now

File: catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java (hunk 11 above)
Comment (pklinef) [reply catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/cql/CqlRequest.java]: I think this will increase the tech debt of this class but since you have already changed it once we can leave it as it is now.
